File: AdventOfCode/2022/11/day11.py
class Monkey:
	ID = 0
	items = []											# items held to be inspected (worry levels stored)
	operation = lambda a : a * 1						# operation to do on inspection
	test = lambda a : True if a % 2 == 0 else False		# test to see worry level
	true = 0											# which monkey ID to pass to if test True
	false = 0											# which monkey ID to pass to if test False
	inspections = 0

	def __init__(self, id):
		self.ID = id
	
	def setItems(self, newItems:list[int]):
		self.items = newItems

	def setOperation(self, newOp):
		self.operation = newOp

	def setTest(self, newTest):
		self.test = newTest

	def setTrue(self, ifTrue):
		self.true = ifTrue

	def setFalse(self, ifFalse):
		self.false = ifFalse

	def inspect(self, idx, part1=False):
		self.items[idx] = int(self.operation(self.items[idx]))
		if part1:
			self.items[idx] = int(self.items[idx] // 3)
		self.inspections += 1

	def check(self, idx) -> int:
		if self.test(self.items[idx]):
			return self.true
		else:
			return self.false
		
class Group:
	monkeys:list[Monkey] = []

	def doRound(self, part1=False):
		for monkey in self.monkeys:
			while len(monkey.items) > 0:
				monkey.inspect(0, part1)
				sendTo = monkey.check(0)
				item = monkey.items.pop(0) # remove from this monkey
				self.monkeys[sendTo].items.append(item) # send to next

	def print(self):
		for monkey in self.monkeys:
			print("Monkey " + str(monkey.ID) + ", inspected " + str(monkey.inspections) + " holds " + str(monkey.items))

	def getMonkeyBusiness(self):
		top = 0
		second = 0
		for monkey in self.monkeys:
			if monkey.inspections > self.monkeys[top].inspections:
				second = top
				top = monkey.ID
			elif monkey.inspections > self.monkeys[second].inspections:
				second = monkey.ID
		topInspections = self.monkeys[top].inspections
		secondInspections = self.monkeys[second].inspections
		business = topInspections * secondInspections
		# print("Top (" + str(top) + "): " + str(topInspections))
		# print("Second (" + str(second) + "): " + str(secondInspections))
		# print("Business:", business)
		return business

	def __init__(self, contents:list[str]):
		currentMonkey = Monkey(0)
		for fullLine in contents:
			# remove ':', ',' and begining whitespace, then split on ' '
			line = fullLine.replace(':', '').replace(',', '').lstrip().split(' ')

			if line[0] == "Monkey":
				currentMonkey = Monkey(int(line[1]))

			elif line[0] == "Starting":
				items = []
				for i in range(2, len(line)):
					items.append(int(line[i]))
				currentMonkey.setItems(items)

			elif line[0] == "Operation":
				op = lambda a : a
				if line[5] == "old":
					if line[4] == '*':
						op = lambda a : a * a
					elif line[4] == '+':
						op = lambda a : a + a
				else:
					b = int(line[5])
					if line[4] == '*':
						op = lambda a, b=b : a * b
					elif line[4] == '+':
						op = lambda a, b=b : a + b
				currentMonkey.setOperation(op)

			elif line[0] == "Test":
				if line[1] == "divisible":
					div = int(line[3])
					currentMonkey.setTest(lambda a, d=div : True if a % d == 0 else False)
					
			elif line[0] == "If":
				if line[1] == "true":
					currentMonkey.setTrue(int(line[-1]))
				if line[1] == "false":
					currentMonkey.setFalse(int(line[-1]))
					self.monkeys.append(currentMonkey)	

# def part1(group:Group):
# 	for _ in range(20):
# 		group.doRound(True)
# 	group.print()
# 	business = group.getMonkeyBusiness()
# 	print("Part 1:", business)

# def part2(group:Group):
# 	for i in range(10000):
# 		print("Round " + str(i) + " " + str())
# 		group.doRound(False)
# 	group.print()
# 	business = group.getMonkeyBusiness()
# 	print("Part 2:", business)

def getBusiness(group:Group, rounds:int, part1=False) -> int:
	for i in range(rounds):
		print("Round " + str(i) + " " + str((i/rounds)*100) + "%")
		group.doRound(part1)
	group.print()
	return group.getMonkeyBusiness()

def main():
	eg1 = "example"
	eg2 = "example2"
	inp = "input"
	filename = inp
	f = open(filename, "r")
	contents = f.read().split("\n")
	
	monkeyGroup = Group(contents)
	# print("Part 1:", getBusiness(monkeyGroup, 20, True))
	print("Part 2:", getBusiness(monkeyGroup, 10000))
	# part1(monkeyGroup)
	# part2(monkeyGroup)

main()

File: AdventOfCode/2022/11/day11.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <queue>
#include <regex>
using namespace std;

class Monkey {
	public:
		int id = 0;
		std::vector<long long> items;
		
		// operation
		bool opOld = false;
		int opMod = 0; // what to +/* by
		bool opMult = false; // + or *

		// test
		int testBy = 1;
		int ifTrue = 0;
		int ifFalse = 0;

		long long inspections = 0;

		string stringItems() {
			string str = "";
			for (int i = 0; i < items.size(); i++) {
				string end = ", ";
				if (i == items.size()-1) {
					end = "";
				}
				str += to_string(items[i]) + end;
			}
			return str;
		}

		string stringOperation() {
			string str = "new = old ";
			if (opMult) {
				str.append("* ");
			} else {
				str.append("+ ");
			}
			if (opOld) {
				str.append("old");
			} else {
				str.append(to_string(opMod));
			}
			return str;
		}

		long long doOp(long long item) {
			if (opOld) {
				if (opMult) {
					return item * item;
				}
				return item + item;
			}
			if (opMult) {
				return item * opMod;
			}
			return item + opMod;
		}

		void inspect(long div, bool part1) {
			++inspections;
			long long item = doOp(items.front());
			if (part1) {
				item /= 3;
			}
			// if (item < 0) {
			// 	cout << "Monkey " << id << " holding " << items.front() << " that became " << item << endl;
			// 	cout << stringOperation() << endl;
			// 	int a = 0/0;
			// }
			items.at(0) = item % div;
		}

		int check() {
			if (items.front() % testBy == 0) {
				return ifTrue;
			}
			return ifFalse;
		}
};

class Group {
	public:
		vector<Monkey> monkeys;
		long div = 1; // avoid overflow by modding by the product of all division checks

		void printDefs() {
			for (auto m : monkeys) {
				cout << "Monkey " << m.id << ":" << endl;
				cout << "  Starting items: " << m.stringItems() << endl;
				cout << "  Operation: " << m.stringOperation() << endl;
				cout << "  Test: divisible by " << m.testBy << endl;
				cout << "    If true: throw to monkey " << m.ifTrue << endl;
				cout << "    If false: throw to monkey " << m.ifFalse << endl;
				cout << endl;
			}
		}

		void print() {
			for (int i = 0; i < monkeys.size(); i++) {
				Monkey m = monkeys.at(i);
				cout << "Monkey " << m.id << ", inspected " << m.inspections << " holds " << m.stringItems() << endl;
			}
		}

		long long getBusiness() {
			long long top = 0;
			long long second = 0;
			for (int i = 0; i < monkeys.size(); i++) {
				long long insp = monkeys[i].inspections;
				if (insp > top) {
					second = top;
					top = insp;
				} else if (insp > second) {
					second = insp;
				}
			}
			long long mult = top * second;
			// cout << top << " * " << second << " = " << mult << endl;
			return mult;
		}

		void doRound(int round, bool part1) {
			for (int i = 0; i < monkeys.size(); ++i) {
				// cout << "\tMonkey " << monkeys[i].id << endl;
				while (monkeys[i].items.size() > 0) {
					// cout << round << " monkey " << monkeys[i].id << ": item " << monkeys[i].items.front() << endl;
					monkeys[i].inspect(div, part1);
					int to = monkeys[i].check();
					monkeys.at(to).items.push_back(monkeys[i].items.front());
					monkeys[i].items.erase(monkeys[i].items.begin());
				}
			}
		}

		string replace(string str, string remove) {
			return std::regex_replace(str, std::regex(remove), "");
		}

		string trimStart(string str) {
			auto begin = str.find_first_not_of(" ");
			return str.substr(begin);
		}

		vector<string> split(string str) {
			vector<string> line;
			string word = "";
			for (auto x : str) {
				if (x == ' ') {
					line.push_back(word);
					word = "";
				} else {
					word += x;
				}
			}
			line.push_back(word);
			return line;
		}

		vector<string> fixLine(string line) {
			vector<string> ret;
			if (line.size() > 0) {
				line = replace(line, ":");
				line = replace(line, ",");
				line = trimStart(line);
				ret = split(line);
			}
			return ret;
		}

		void read(ifstream &in) {
			string startLine = "";
			Monkey curr;
			while (getline(in, startLine)) {
				vector<string> line = fixLine(startLine);
				if (line.size() == 0) {
					continue;
				}
				
				if (line.at(0) == "Monkey") {

					curr.id = stoi(line.at(1));

				} else if (line.at(0) == "Starting") {

					vector<long long> items;
					for (int i = 2; i < line.size(); i++) {
						items.push_back(stoi(line.at(i)));
					}
					curr.items = items;

 				} else if (line.at(0) == "Operation") {

					curr.opMult = line.at(4) == "*";
					curr.opOld = line.at(5) == "old";
					if (!curr.opOld) {
						curr.opMod = stoi(line.at(5));
					}
					
				} else if (line.at(0) == "Test") {

					curr.testBy = stoi(line.at(3));
					div *= curr.testBy;
					
				} else if (line.at(0) == "If") {
					
					if (line.at(1) == "true") {
						curr.ifTrue = stoi(line.at(line.size()-1));
					} else {
						curr.ifFalse = stoi(line.at(line.size()-1));
						monkeys.push_back(curr);
					}
				}
			}
		}
};

long long getBusiness(Group group, int rounds, bool part1) {
	float r = rounds;
	for (int i = 1; i <= rounds; ++i) {
		// cout << "Round " << i << ": " << ((i/r)*100) << "%" << endl;
		group.doRound(i, part1);
		if (i == 1 || i == 20 || i%1000 == 0) {
			cout << "== After round " << i << " ==" << endl;
			group.print();
		}
	}
	// group.print();
	return group.getBusiness();
}

int main() {
	string filename = "input";
	ifstream in (filename);
	
	if (in.is_open()) {
		Group group;
		group.read(in);
		cout << group.div << endl;
		// group.printDefs();
		// group.print();
		// cout << "Business (1): " << getBusiness(group, 20, true) << endl;
		// cout << "Business (2): " << getBusiness(group, 10000, false) << endl;
	}
}